crypto/evp_extra/p_pqdsa_asn1.c

```diff
@@ -149,7 +149,40 @@ static int pqdsa_priv_decode(EVP_PKEY *out, CBS *params, CBS *key, CBS *pubkey)
     OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
     return 0;
   }
-  return PQDSA_KEY_set_raw_private_key(out->pkey.pqdsa_key,CBS_data(key));
+
+  // Set the private key
+  if (!PQDSA_KEY_set_raw_private_key(out->pkey.pqdsa_key, CBS_data(key))) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  // Calculate public key from private key if method exists
+  if (out->pkey.pqdsa_key->pqdsa->method->pqdsa_pack_pk_from_sk != NULL) {
+    size_t pk_len = out->pkey.pqdsa_key->pqdsa->public_key_len;
+    uint8_t *public_key = OPENSSL_malloc(pk_len);
+
+    if (public_key == NULL) {
+      OPENSSL_PUT_ERROR(EVP, ERR_R_MALLOC_FAILURE);
+      return 0;
+    }
+    // attempt to construct the public key from priv
+    if (!out->pkey.pqdsa_key->pqdsa->method->pqdsa_pack_pk_from_sk(
+            public_key, CBS_data(key))) {
+      OPENSSL_free(public_key);
+      OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+      return 0;
+    }
+    // set the public key
+    int ret = PQDSA_KEY_set_raw_public_key(out->pkey.pqdsa_key, public_key);
+    OPENSSL_free(public_key);
```

Comment:
Since public_key was allocated internally can the PKEY just take ownership of it without having to copy it?

Reply:
I don't see this as being possible, as the size of the `public_key` is dynamic, and defined by `out->pkey.pqdsa_key->pqdsa->public_key_len`. Am I missing something?

Reply:
Yeah, you allocate the key on L162, looking at `PQDSA_KEY_set_raw_public_key` it just calls `OPENSSL_memdup` which allocates a new buffer and memcpy the data into that new buffer, then this function frees `public_key`. Why not store `public_key` in `out->pkey.pqdsa_key->public_key`?

Reply:
got it! [54224d2](https://github.com/aws/aws-lc/pull/2145/commits/54224d23a3af931471603b35e308d46f3e9e0349)